docs/cvl/hooks.md

```diff
@@ -20,77 +22,422 @@ Syntax
+`currentContract`.
+```
+
+```{todo}
+Does the contract need to be the contract that defines the field, or can it be
+an inheriting contract?  What happens if there are multiple variables with the
+same name (because of inheritance)?
+```
+
+[storage-layout]: https://docs.soliditylang.org/en/v0.8.17/internals/layout_in_storage.html
+
+If the indicated location holds a struct, you can refer to a specific field of
+the struct by appending `.<field-name>` to the path.  For example, the following
+hook will execute on every store to the `balance` field of the struct `C.owner`:
+```cvl
+hook Sstore C.owner.balance uint b STORAGE { ... }
+```
+
+If the indicated location holds an array, you can refer to an arbitrary element
+of the array by appending `[INDEX uint <variable>]`.  This pattern will match
+any store to an element of the array, and will bind the named variable to the
+index of the access.  For example, the following hook will execute on any write
+to the array `C.entries` and will update the corresponding entry of the ghost
+mapping `_entries` to match:
+```cvl
+hook Sstore C.entries[INDEX uint i] uint e STORAGE {
+    _entries[i] = e;
+}
+```
+
+Similarly, if the indicated location holds a mapping, you can refer to an
+arbitrary entry by appending `[KEY <type> <variable>]`.  This pattern will
+match any write to the mapping, and will bind the named variable to the key.
+For example, the following hook will execute on any write to the mapping
+`C.balances`, and will update the `_balances` ghost accordingly:
+
+```cvl
+hook Sstore C.balances[KEY address user] uint balance STORAGE {
+    _balances[user] = balance;
+}
+```
+
+Finally, there is the low-level access pattern `<base>.(offset <n>)` for matching
+loads and stores that are a specific number of bytes from the
+base.  For example, the following hook will match writes to the third or fourth
+byte of slot 1 (these two bytes are matched because the type of the variable is
+`uint16`:
+
+```cvl
+hook Sstore (slot 1).(offset 2) uint16 b STORAGE { ... }
+```
+
+These different kinds of paths can be combined.  For example, the following
+hook will execute whenever the contract writes to the `balance` field of a
+struct in the `users` mapping of contract `C`:
+```cvl
+hook C.users[KEY address user].balance uint v (uint old_value) STORAGE { ... }
+```
+Inside the body of the hook, the variable `user` will refer to the address that
+was used as the key into the mapping `C.users`; the variable `v` will contain
+the value that is written, and the variable `old_value` will contain the value
+that was previously stored there.
+
+```{note}
+The only available access paths for `solc` versions 5.17 and older are `slot`
+and `offset` paths.
+```
+
+### Access path caveats
+
+```{todo}
+information on what kind of funny business happens if people do their own memory
+management, and about the disjointness of arrays and so forth
+
+How can you tell if an analysis failed and your hooks didn't apply?
```

Comment:
maybe suffices to say that if the relevant analysis fails and hooks cannot be applied, the rule will fail with a suitable message and an error will show in the errors tab of the web report